File: python/ejercicio1.py
# Made by Jose Gutierrez (@jg2kpy) https://github.com/jg2kpy https://juniorgutierrez.com/

# Importo librerias
import re  # Libreria de regex
import requests  # Libreria para realizar peticiones HTTP
from bs4 import BeautifulSoup  # BeautifulSoup4 Libreria para analizar HTML
import matplotlib.pyplot as plt  # Libreria para graficar
import time # Libreria para generar un cooldown

# Funcion principal
def main():
    # Lista de los 20 lenguajes mas usados segun tiobe index
    tiobe_index = [['Python', 0, 0], ['C', 0, 0], ['Java', 0, 0], ['Cpp', 0, 0], ['Csharp', 0, 0], ['VisualBasic', 0, 0], ['JavaScript', 0, 0], ['Assembly language', 0, 0], ['SQL', 0, 0], ['PHP', 0, 0], ['R', 0, 0], ['Delphi', 0, 0], ['Go', 0, 0], ['Swift', 0, 0], ['Ruby', 0, 0], ['visual-basic-6', 0, 0], ['Objective-C', 0, 0], ['Perl', 0, 0], ['Lua', 0, 0], ['matlab', 0, 0]]

    MAX = 0
    MIN = 0
    try:  # Empezamos el Web Scrapping
        print('Realizando WebScrapping a github.com y escribiendo resultados en Resultados.txt')
        print('Esta operacion puede tomar aproximadamente 1 minuto...')
        for language in tiobe_index:  # Iteramos la lista para obtener la cantidad de repositorios de github por cada topico
            tries = 0
            print(f'Scrapping...{language[0]}',end='')
            url = 'https://www.github.com/topics/' + language[0]
            response = requests.get(url)  # Solicitud HTTP a github.com/topics
            responseHTML = BeautifulSoup(response.text, features='html.parser')
            # Usamos el atributo HTML class para obtener la informacion que queremos
            match_topic = responseHTML.find(class_='h3 color-fg-muted')
            # Si no obtenemos la respuesta probamos otra 3 veces, si aun no tenemos respuesta se saca ese lenguaje de la lista
            while match_topic == None and tries < 3:
                tries = tries + 1
                time.sleep(2) # Se espera 2 segundos como cooldown entre cada intento
                response = requests.get(url)
                responseHTML = BeautifulSoup(response.text, features='html.parser')
                match_topic = responseHTML.find(class_='h3 color-fg-muted')

            if match_topic == None:
                print(f'\nError al obtener el lenguaje {language[0]}')
                tiobe_index.pop(tiobe_index.index(language))
            else:
                match_topic = match_topic.text.replace(',', '')
                # Usamos regex para obtener las apariciones
                language[1] = int(re.findall('[0-9]+', match_topic)[0])

                print(f': {language[1]}')
                # Guardamos los resultado en el archivo Resultados.txt
                with open('Resultados.txt', "a") as f:
                    f.write(f'{language[0]},{language[1]}\n')

                # Actualizamos el maximo y el minimo en cada iteracion
                if MAX < language[1]:
                    MAX = language[1]

                if MIN > language[1] or MIN == 0:
                    MIN = language[1]

    except requests.exceptions.RequestException:
        print('Error al cargar la pagina de github.com')
        return -1

    dif = MAX - MIN
    for language in tiobe_index:  # Aplicamos la formula de GitHub Rating
        language[2] = ((language[1] - MIN)/dif) * 100

    tiobe_index.sort(key=mySort)  # Ordenamos la lista

    for language in tiobe_index:  # Mostramos el resultado en Pantalla
        print(f"{language[0]}, {language[1]}, {language[2]}")

    # Graficamos mediante matplotlib
    left = []
    height = []
    tick_label = []

    n = len(tiobe_index)
    for i in range(1, 11):
        left.append(i)
        height.append(tiobe_index[n-i][1])
        tick_label.append(tiobe_index[n-i][0])

    plt.bar(left, height, tick_label=tick_label)

    plt.xlabel('NOMBRE_LENGUAJE')
    plt.ylabel('NRO_APARICIONES')
    plt.title('GitHub')

    plt.show()


def mySort(e):  # Funcion para que sort sepa por cual parametro ordenar
    return e[2]

# Con ese if verificamos que esta archivo esta siendo ejecutado y no importado como una libreria
if __name__ == '__main__':
    main()

File: php/ejercicio1.php
<?php
set_time_limit(0); // para poder utilizar sleep
#variables
$listaLenguajes = [
	["lenguaje"=>"python", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"ruby", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"c", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"cpp", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"csharp", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"objective-c", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"visualbasic", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"visual-b", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"java", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"javascript", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"assembly-language", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"sql", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"php", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"r", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"delphi", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"go", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"swift", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"perl", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"lua", "apariciones"=>0, "rating"=>0 ],
	["lenguaje"=>"matlab", "apariciones"=>0, "rating"=>0 ]
];
$ch = curl_init();
$min = PHP_INT_MAX;
$max = PHP_INT_MIN;
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
#webscraping
foreach ($listaLenguajes as $posicion => $lengActual) {
	$intentos = 0;
	$url = "https://github.com/topics/".$lengActual["lenguaje"];
	curl_setopt($ch, CURLOPT_URL, $url);
	// intentar 3 veces en caso de error
	while ($intentos < 3) {
		$html = curl_exec($ch);
		$httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
		$intentos++;
		if ($httpCode == 200) {
			$intentos = 0;
			break;
		}
		usleep(1000000); // esperar 1s si ocurre error
	}
	#parsing
	if ($intentos == 0) {
		$re = '/<h2 class="h3 color-fg-muted">\n.*\n.*\n.*\n.*<\/h2>/m';
		preg_match_all($re, $html, $matches, PREG_SET_ORDER, 0);
		$re = '/[^0-9]+/';
		$string = strip_tags($matches[0][0]);
		$apariciones = preg_replace($re, "", $string);
		$listaLenguajes[$posicion]["apariciones"] = $apariciones;
		$min = min($min, $apariciones);
		$max = max($max, $apariciones);
	}
	#error
	else {
		print($lengActual["lenguaje"]." ignorado<br>");
		unset($listaLenguajes[$posicion]); // quitar de la lista
	}
}
curl_close($ch);
escribirArchivo();
obtenerRatingOrdenado($min, $max);
#imprimir en pantalla
foreach ($listaLenguajes as $lengActual) {
	print($lengActual["lenguaje"].", ".round($lengActual["rating"], 3).", ".$lengActual["apariciones"]."<br>");
}
print("<br>");
obtenerListaGrafico();

#funciones
function escribirArchivo () {
	global $listaLenguajes;
	$archivo = fopen("resultados.txt", "w");
	#escribir en el archivo
	foreach ($listaLenguajes as $posicion => $lengActual) {
		fwrite($archivo, $lengActual["lenguaje"].", ".$lengActual["apariciones"]."\n");
	}
	fclose($archivo);
}

#se obtiene el rating de cada lenguaje y se ordena descendentemente
function obtenerRatingOrdenado($min, $max) {
	global $listaLenguajes;
	#calcular rating
	foreach ($listaLenguajes as $posicion => $lengActual) {
		$listaLenguajes[$posicion]["rating"] = ($lengActual["apariciones"] - $min)/($max - $min)*100;
	}
	#ordenar descendentemente por rating
	uasort($listaLenguajes, function($a, $b) {
		return $b["rating"] > $a["rating"];
	});
}

#corta la lista al top 10 y transforma al formato del grafico
function obtenerListaGrafico() {
	global $listaLenguajes;
	#preparar datos para el grafico
	$listaLenguajes = array_slice($listaLenguajes, 0, 10); // se agarran los 10 primeros
	$listaLenguajes = array_map(function($listaLenguajes) { // se modifican las claves del array para graficar
		return array(
		'label' => $listaLenguajes["lenguaje"],
		'y' => $listaLenguajes["apariciones"]
		);
	}, $listaLenguajes);
}
?>

<!-- grafico con canvasjs -->
<!DOCTYPE HTML>
<html>
<head>
<script>
window.onload = function() {

	var chart = new CanvasJS.Chart("chartContainer",
	{
		animationEnabled: true,
		axisY: {
			title: "Apariciones"
		},
		axisX: {
			title: "Lenguajes"
		},
		data: [{
			type: "column",
			dataPoints: <?php echo json_encode($listaLenguajes, JSON_NUMERIC_CHECK); ?>
		}]
	});
	chart.render();
}
</script>
</head>
<body>
<div id="chartContainer" style="height: 370px; width: 100%;"></div>
<script src="https://canvasjs.com/assets/script/canvasjs.min.js"></script>
</body>
</html>